### PR TITLE
`ukernel.generic_raw` op passing only buffer and offset for memref operands

### PR DESCRIPTION
**Pull request by bjacob**
This is less opinionated than the existing `ukernel.generic` in that `memref` operands are lowered to just base buffer and offset operands, the bare minimum for any callee to be able to do anything with the memref.

By not appending stride values, this leaves it up to each use case to decide which strides to pass.

This makes `ukernel.generic_raw` flexible enough to replace all other existing ukernel ops. Though we want to retain `ukernel.generic` for now, we will at least proceed with migrating `ukernel.mmt4d` and all the ukernels in VMVXOps.td.

This PR also does some naming harmonization: `micro_kernel` -> `ukernel`, `foo_ukernel` -> `ukernel_foo`.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/test/lower_ukernel_to_calls.mlir [hunk 1]
```diff
@@ -81,7 +81,88 @@ func.func @generic_ukernel(%arg0 : memref<?x?xf32>, %arg1 : memref<?x?xf32>,
...
+  return
+}
+//      CHECK: func.func private @scalar_fn(i32, f64, index, memref<f32>)
+//      CHECK: func.func @ukernel_generic_raw_scalar_types
+// CHECK-SAME:     %[[ARG0:[a-zA-Z0-9]+]]: i32
+// CHECK-SAME:     %[[ARG1:[a-zA-Z0-9]+]]: f64
+// CHECK-SAME:     %[[ARG2:[a-zA-Z0-9]+]]: index
+// CHECK-SAME:     %[[ARG3:[a-zA-Z0-9]+]]: memref<f32>
+//      CHECK:   %[[BASE:.+]], %[[OFFSET:.+]] = memref.extract_strided_metadata %[[ARG3]]
+//      CHECK:   call @scalar_fn(%[[ARG0]], %[[ARG1]], %[[ARG2]], %[[BASE]])
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/test/lower_ukernel_to_calls.mlir]: I think this needs to include offset too.

File: compiler/src/iree/compiler/Codegen/Dialect/UKernelOps.cpp [hunk 2]
```diff
@@ -127,52 +138,90 @@ static LogicalResult lowerToCallOperands(Location loc, RewriterBase &rewriter,
...
       })
       .Default([](Type) { return failure(); });
 }
 
-FailureOr<func::CallOp> UKernelGenericOp::lowerToFunctionCall(
-    RewriterBase &rewriter) {
+/// Shared implementation of
+///   {UKernelGenericOp, UKernelGenericRawOp}::lowerToFunctionCall
+/// which differ only in the bufferPassingStyle.
+FailureOr<func::CallOp> lowerUKernelGenericToFunctionCall(
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Dialect/UKernelOps.cpp]: Make this `static`?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Mostly looks good! Thanks for doing this. The only issue here is that `ukernel.generic` and `ukernel.generic_raw` need to lower `memref<f32>` to two operands in the `func.func` and `func.call` , the pointer and the offset. 
It is a bit confusing that the `memref<f32>` in a `fn.call` lowers to a `void *`, but thats the path that works today...

File: compiler/src/iree/compiler/Codegen/Common/test/lower_ukernel_to_calls.mlir (hunk 1 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Common/test/lower_ukernel_to_calls.mlir]: Right, this is a preexisting issue and being addressed in #13054.

Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Common/test/lower_ukernel_to_calls.mlir]: As discussed on Discord, updated this PR to add the commit from #13054.

File: compiler/src/iree/compiler/Codegen/Dialect/UKernelOps.cpp [hunk 3]
```diff
@@ -65,19 +69,22 @@ static FailureOr<func::CallOp> createFunctionCall(RewriterBase &rewriter,
...
-}
+/// Enum to describe what function arguments to pass to the ukernel target
+/// function for each memref operand of the ukernel op.
+enum class BufferPassingStyle {
+  // Pass only the base buffer and offset, the bare minimum for the callee to be
+  // able to compute the effective buffer address.
+  kBufferAndOffset,
+  // Also pass the N-1 first strides for a rank-N memref, i.e. all but the
+  // inner-most stride.
+  kBufferAndOffsetAndNm1Strides
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Dialect/UKernelOps.cpp:85]: Please add a TODO (and if you feel like it file a bug as well and assign to me, but TODO with my name here is enough) to drop this to not use `OffsetAndNm1Strides`, but instead use `OffsetSizesAndStrides`.

**Review by MaheshRavishankar — APPROVED** [review]
Nice! Thanks!

File: compiler/src/iree/compiler/Codegen/Dialect/UKernelOps.cpp (hunk 3 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Dialect/UKernelOps.cpp:85]: Done!

**Pull request merged**